File: bot/src/main/kotlin/org/aeriagloris/telegram/commands/WhoisCommand.kt
package org.aeriagloris.telegram.commands

import org.telegram.telegrambots.api.methods.send.SendMessage
import org.telegram.telegrambots.api.objects.Chat
import org.telegram.telegrambots.api.objects.User
import org.telegram.telegrambots.bots.AbsSender
import org.telegram.telegrambots.logging.BotLogger
import org.aeriagloris.persistence.JdbcStore
import org.jetbrains.exposed.sql.transactions.transaction
import org.aeriagloris.persistence.schema.*
import org.jetbrains.exposed.sql.*

class WhoisCommand(val store: JdbcStore) : ExtendedCommand("whois", "Query telegram or PSN id")
{
    override fun execute(absSender: AbsSender, user: User, chat: Chat, arguments: Array<String>)
    {
        if (arguments.size != 1) {
            sendReply(absSender, chat, "To query user provide his @TelegramId (starting with @) or PsnId")
            return
        }

        transaction {
            logger.addLogger(StdOutSqlLogger())

            val dbUser = Guardian.find { Guardians.telegramName eq user.getUserName() }.singleOrNull()

            if (dbUser == null) {
                sendReply(absSender, chat, "You need to link your PSN account first: use /psn command")
            } else {
                val name = arguments[0]

                val guardian = if (name.startsWith("@")) {
                        Guardian.find { Guardians.telegramName eq name.drop(1) }.singleOrNull()
                    } else {
                        Guardian.find { Guardians.psnName eq name }.singleOrNull()
                    }

                if (guardian == null) {
                    sendReply(absSender, chat, "Guardian "+name+" was not found.")
                } else {
                    sendReply(absSender, chat, "Guardian @"+guardian.telegramName+" PSN "+guardian.psnName)
                }
            }
        }
    }
}

File: src/commands/whois_command.rs
use crate::{commands::validate_username, commands::guardian_lookup, models::Guardian};
use crate::{Bot, BotCommand, DbConnection};
use diesel::prelude::*;
use futures::Future;

pub struct WhoisCommand;

command_ctor!(WhoisCommand);

impl BotCommand for WhoisCommand {
    fn prefix(&self) -> &'static str {
        "/whois"
    }

    fn description(&self) -> &'static str {
        "Query telegram or PSN id"
    }

    fn execute(
        &self,
        bot: &Bot,
        message: &telebot::objects::Message,
        _command: Option<String>,
        name: Option<String>,
    ) {
        if name.is_none() {
            return bot.send_plain_reply(
                &message,
                "To query user provide his @TelegramId (starting with @) or PsnId".into(),
            );
        }

        let name = name.unwrap();
        let connection = bot.connection();

        if validate_username(bot, &message, &connection).is_none() {
            return;
        }

        let guardian = guardian_lookup(&name, &connection);

        match guardian {
            Ok(Some(guardian)) => {
                bot.send_plain_reply(
                    &message,
                    format!(
                        "Guardian @{telegram_name} PSN {psn_name}",
                        telegram_name = guardian.telegram_name,
                        psn_name = guardian.psn_name
                    ),
                );
            }
            Ok(None) => {
                bot.send_plain_reply(&message, format!("Guardian {} was not found.", name));
            }
            Err(_) => {
                bot.send_plain_reply(&message, "Error querying guardian name.".into());
            }
        }
    }
}
